### ISSUE TITLE
Config option to reference private registry

### ISSUE THREAD

**Issue by chalky9909** [heptio/sonobuoy#160]
We are attempting to deploy sonobuoy on a k8s cluster configured behind a firewall with no internet access.

The sonobuoy images are available in a private registry and quickstart.yaml updated to reflect this.  When launching the deployment, the sonobuoy specific pods launch with no issues.

However, after a period of time, the below pods fails to launch in the newly created "e2e-tests" namespace.

A kubectl describe against this:
`kubectl describe pod pod-submit-remove-16bb7011-d055-11e7-ba38-0a580a02306e -n=e2e-tests-pods-jgshk`

shows that it is failing to pull an nginx image as below:

`Failed to pull image "gcr.io/google-containers/nginx-slim-amd64:0.20": rpc error: code = Unknown desc = Error response from daemon: Get https://gcr.io/v1/_ping: dial tcp 173.194.69.82:443: i/o timeout`

As the cluster does not have internet access, it is unable to pull and access this image from gcr.io

This nginx image from gcr.io is not referenced in the deployment yaml so is it possible to add a configuration option where the nginx image can be accessed from a private internal registry?

Something like the below if this is feasible:
`private-registry:443/gcr.io/google-containers/nginx-slim-amd64:0.20`

Any assistance with this would be much appreciated.

**Comment by timothysc** [comment]
It's been on the radar for a while to help support air-gapped environments.  The problem is sifting through all the images required to build and adding an option for the generated .yaml.

**Comment by chalky9909** [comment]
@timothysc thanks for taking a look at this request. I take it then that it is not just the image that is being referred to as below and there are others that are required:

Failed to pull image "gcr.io/google-containers/nginx-slim-amd64:0.20"

**Comment by timothysc** [comment]
@chalky9909 - There are several different images.

**Comment by liztio** [comment]
@chalky9909 I'm working on some changes to the the tests to make it feasible to run them on airgapped clusters. Unfortunately it requires upstream changes, and everything is currently in lockdown for the 1.9 release. Once that clears, I should be able to get you an update.

Watch this space!

**Comment by chalky9909** [comment]
@liztio thanks for the update.  Will be happy to test once the upstream changes have been made.

**Comment by erwbgy** [comment]
@liztio can you provide a link to the upstream issue?

**Comment by mweigel** [comment]
@liztio I'd also be happy to test this once ready, thanks.

**Comment by lewismarshall** [comment]
@liztio I've needed to run this myself and I'de started on a PR and didn't see your comments. Is it too late for me to share my WIP for sonobouy and the upstream e2e.test binary?

Also I'm having trouble finding the source for the image gcr.io/heptio-images/heptio-e2e:master

**Comment by timothysc** [comment]
@lewismarshall We have not yet open-sourced our extended tests 

/cc @kstewart @jbeda

**Comment by chuckha** [comment]
Are we punting this out of the milestone or is there something actionable we can do here?

**Comment by timothysc** [comment]
Given the state of upstream I'm fine with punting to 1.0

**Comment by timothysc** [comment]
Dropping this breadcrumb here when we loop back to this - kubernetes/kubernetes#38067

**Comment by zbindenren** [comment]
This feature got [merged](https://github.com/kubernetes/kubernetes/pull/60848/commits) in master.

It is now possible to reference a yaml file with `KUBE_TEST_REPO_LIST` environment variable. In this file it is possible to override the environment registries:

```yaml
dockerLibraryRegistry: docker.io/library
e2eRegistry: gcr.io/kubernetes-e2e-test-images
gcRegistry: k8s.gcr.io
privateRegistry: gcr.io/k8s-authenticated-test
sampleRegistry: gcr.io/google-samples
```

Since 0.12 is already released, I wonder if this feature still coming to 0.12.x?

**Comment by timothysc** [comment]
@zbindenren the ability to obtain the manifests exists, but the plumbing to override through the e2e tests upstreams for all the different images does not.  I will open an issue to cross reference.

**Comment by ashishapy** [comment]
This issue is blocker for any deployment in corporate env. I guess, it is asking a lot from a open source project, any ETA on this.
Thanks

**Comment by xdrus** [comment]
Would it be possible to get a list of containers required for e2e tests?

It can unblock us because we can fetch them trough a corporate artifactory docker-registry proxy and re-tag to original names.

**Comment by joshrosso** [comment]
@timothysc or @liztio, is what's below an accurate summary of this issue's status?

1. With https://github.com/kubernetes/kubernetes/pull/60848 in `13.0`+, using private repositories in E2E tests **is** possible.

1. We need to update the plumbing between Sonobuoy and the E2E tests to resolve this issue (https://github.com/heptio/sonobuoy/issues/160).

1.  Until then, there are 2 workarounds.

    3a. Pre-load images needed by E2E tests on **all** worker nodes.

    3b. Write a MutatingWebhook admission controller that alters the `image:` field on E2E-test pods to use a private repository.

Thanks in advanced!

**Comment by erwbgy** [comment]
We successfully use 3b on-prem.

**Comment by johnSchnake** [comment]
@joshrosso That seems to be my understanding but I haven't (yet) put it to the test. I'm going to see if I can create an airgapped cluster/private registry and really get tests to run that way. I wonder if we are going to find that there are some tests that still reference images directly in the default registries.

**Comment by joshrosso** [comment]
@johnSchnake, i did an airgapped run using **3a** on 1.13.3.

There were a few tests that referenced `nginx:latest` and 2 that had `imagePullPolicy: Always`. The latter was the more problematic (as a retag of the versioned nginx image to latest worked around the former).

Hope this helps if you go down a similar path.

**Comment by stevesloka** [comment]
xref #593

**Comment by johnSchnake** [comment]
Working on the solution based on (1) above. I'll put up a PR soon. Manually I was able to just update the yaml to create a configmap with the repo information, mount it into the kube-conformance image, and set an env var to tell k8s to use it.

My test was technically against a _public_ (but custom) registry which seems to be the bigger concern to get airgap to work. Not sure if pulling from a private registry (auth required) going to be different or a hard requirement for this ticket. IMO they seem like they are separate concerns.

**Comment by stevesloka** [comment]
I think a public registry is fine. Did you hack up your /etc/hosts to make sure that docker.io & gcr.io weren't getting called? I guess you could look at the images on a node when the tests are complete to verify as well. I have some PR's coming that will help with testing this process (e.g. image list, image load, etc).

**Comment by johnSchnake** [comment]
The latter is what I did, just verifying the images that were pulled. I first actually tested against an empty registry to confirm it wouldn't work (and the test container was failing due to the image pull issue) and then pushed the image up and watched the test pass.

The reason I did it this way was so that I could use `docker.io` but my own custom images (so I was actually using `docker.io/schnake` instead of `/library`. This drastically simplified the setup since I was still having problems (and debated the value of) setting up my own public registry which I could get my KIND cluster to trust and which I could load with images.
